src/graph_algos.rs

```diff
@@ -1,9 +1,98 @@
-use crate::graph::{ PageGraph, Edge, EdgeId, Node, NodeId };
+use crate::graph::{ PageGraph, Edge, EdgeId, Node, NodeId, FrameId };
 use crate::types::{ EdgeType, NodeType };
 
 use petgraph::Direction;
 
 impl PageGraph {
+    pub fn all_remote_frame_ids(&self) -> Vec<FrameId> {
+        self.nodes.iter().filter_map(|(_node_id, node)|
+            if let NodeType::RemoteFrame { frame_id } = node.node_type {
+                Some(frame_id)
+            } else {
+                None
+            }
+        ).collect()
+    }
+
+    /// Inserts the graph for a given frame into this graph, namespacing ids to avoid conflicts.
+    /// The matching `remote frame` node will gain two new outgoing `cross DOM` edges to the `DOM
+    /// root` and `parser` nodes from the frame.
+    pub fn merge_frame(&mut self, frame_graph: PageGraph, frame_id: &FrameId) {
+        assert!(self.desc.is_root);
+        assert!(!frame_graph.desc.is_root);
+
+        // Find the single `remote frame` node with the specified `frame_id`
+        let matching_remote_frames = self.filter_nodes(|n| if let NodeType::RemoteFrame { frame_id: node_frame_id } = n { node_frame_id == frame_id } else { false });
+        assert!(matching_remote_frames.len() == 1);
+        let remote_frame = matching_remote_frames[0].0.clone();
+
+        // Find the frame's single "DOM root" node with no incoming "cross DOM" edges
+        let matching_dom_roots: Vec<_> = frame_graph.nodes.iter().filter(|(id, node)| {
+            if let NodeType::DomRoot { .. } = node.node_type {
+                frame_graph.graph.edges_directed(**id, Direction::Incoming).map(|(_, _, edge_ids)| {
+                    edge_ids.iter().filter(|edge_id| {
+                        if let EdgeType::CrossDom {} = frame_graph.edges.get(&edge_id).unwrap().edge_type { true } else { false }
+                    }).collect::<Vec<_>>()
+                }).flatten().collect::<Vec<_>>().is_empty()
+            } else {
+                false
+            }
+        }).collect();
+        assert!(matching_dom_roots.len() == 1);
```

Comment:
I added this assertion to be sure, but so far it looks like there aren't any graphs with more than one top-level `DOM root` node.